odiglet/pkg/kube/runtime_details/pods_controller.go

```diff
@@ -0,0 +1,153 @@
+	// We care about pod once it's status changes from not running to running -
+	// this is the point in time we want to apply runtime details detection.
+	oldPod, oldOk := e.ObjectOld.(*corev1.Pod)
+	newPod, newOk := e.ObjectNew.(*corev1.Pod)
+
+	if !oldOk || !newOk {
+		return false
+	}
+
+	return oldPod.Status.Phase != corev1.PodRunning && newPod.Status.Phase == corev1.PodRunning
+}
+
+func (p *podPredicate) Delete(e event.DeleteEvent) bool {
+	// no runtime details detection needed when a pod is deleted
+	return false
+}
+
+func (p *podPredicate) Generic(e event.GenericEvent) bool {
+	// no runtime details detection needed for generic events
+	return false
+}
+
+type PodsReconciler struct {
+	client.Client
+	Scheme    *runtime.Scheme
+	Clientset *kubernetes.Clientset
+}
+
+// We need to apply runtime details detection for a new running pod in the following cases:
+// 1. When a new workload generation is applied, the runtime details might be changed (different env, versions, etc).
+// 2. When a source is added, but there are no running pods yet. When the first pod starts running, this is chance to apply runtime details detection.
+func (p *PodsReconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	logger := log.FromContext(ctx)
+
+	var pod corev1.Pod
+	err := p.Client.Get(ctx, request.NamespacedName, &pod)
+	if err != nil {
+		return reconcile.Result{}, client.IgnoreNotFound(err)
+	}
+
+	podWorkload, err := p.getPodWorkloadObject(ctx, &pod)
+	if err != nil {
+		logger.Error(err, "error getting pod workload object")
+		return reconcile.Result{}, err
+	}
+	if podWorkload == nil {
+		// pod is not managed by a workload, no runtime details detection needed
+		return reconcile.Result{}, nil
+	}
+
+	// get instrumentation config for the pod to check if it is instrumented or not
+	instrumentationConfigName := workload.CalculateWorkloadRuntimeObjectName(podWorkload.Name, podWorkload.Kind)
+	instrumentationConfig := odigosv1.InstrumentationConfig{}
+	err = p.Client.Get(ctx, client.ObjectKey{Name: instrumentationConfigName, Namespace: podWorkload.Namespace}, &instrumentationConfig)
+	if err != nil {
+		return reconcile.Result{}, client.IgnoreNotFound(err)
+	}
+
+	podGeneration, err := GetPodGeneration(ctx, p.Clientset, &pod)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// prevent runtime inspection on pods for which we already have the runtime details for this generation
+	if podGeneration == 0 || podGeneration <= instrumentationConfig.Status.ObservedWorkloadGeneration {
+		logger.Info("skipping redundant runtime details detection since generation is not newer", "name", request.Name, "namespace", request.Namespace, "currentPodGeneration", podGeneration, "observedWorkloadGeneration", instrumentationConfig.Status.ObservedWorkloadGeneration)
+		return reconcile.Result{}, nil
+	}
+
+	// The reconcile should be called every time the pod is updated and changes from not running to running
```

Comment:
This comment seems not related to the code here
```suggestion
	// Perform runtime inspection once we know the pod is newer that the latest runtime inspection performed and saved.
```